spark-3.4/clickhouse-spark/src/main/scala/xenon/clickhouse/func/clickhouse/cityhash/CityHash_v1_0_2.java

```diff
@@ -0,0 +1,344 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenon.clickhouse.func.clickhouse.cityhash;
+
+import io.netty.buffer.ByteBuf;
+
+// copy from https://github.com/dpoluyanov/achord/blob/master/src/main/java/com/github/mangelion/achord/CityHash_v1_0_2.java
+// fixed some bugs involving int32 to uint32 conversion
+final public class CityHash_v1_0_2 {
+
+    private static final long kMul = 0x9ddfea08eb382d69L;
+    // Some primes between 2^63 and 2^64 for various uses.
+    private static final long k0 = 0xc3a5c85c97cb3127L;
+    private static final long k1 = 0xb492b66fbe98f273L;
+    private static final long k2 = 0x9ae16a3b2f90404fL;
+    private static final long k3 = 0xc949d7c7509e6557L;
+
+    private CityHash_v1_0_2() { /* restricted */ }
+
+    private static long Fetch64(ByteBuf p, int index) {
+        return p.getLongLE(index);
+    }
+
+    private static int Fetch32(ByteBuf p, int index) {
+        return p.getIntLE(index);
+    }
+
+    private static long toUint32(int x) {
+        return x & 0xFFFFFFFFL;
+    }
+
+    // Equivalent to Rotate(), but requires the second arg to be non-zero.
+// On x86-64, and probably others, it's possible for this to compile
+// to a single instruction if both args are already in registers.
+    private static long RotateByAtLeast1(long val, int shift) {
+        return (val >>> shift) | (val << (64 - shift));
+    }
+
+    private static long ShiftMix(long val) {
+        return val ^ (val >>> 47);
+    }
+
+    private static long Uint128Low64(UInt128 x) {
+        return x.first;
+    }
+
+    private static long Rotate(long val, int shift) {
+        return shift == 0 ? val : (val >>> shift) | (val << (64 - shift));
+    }
+
+    private static long Uint128High64(UInt128 x) {
+        return x.second;
+    }
+
+    // Hash 128 input bits down to 64 bits of output.
+// This is intended to be a reasonably good hash function.
+    public static long Hash128to64(UInt128 x) {
+        // Murmur-inspired hashing.
+        long a = (Uint128Low64(x) ^ Uint128High64(x)) * kMul;
+        a ^= (a >>> 47);
+        long b = (Uint128High64(x) ^ a) * kMul;
+        b ^= (b >>> 47);
+        b *= kMul;
+        return b;
+    }
+
+    private static long HashLen16(long u, long v) {
+        return Hash128to64(UInt128.of(u, v));
+    }
+
+    private static long HashLen0to16(ByteBuf s, int index, int len) {
```

Comment:
We can totally avoid `ByteBuf` in this class, as the `clickhouse-java` does

https://github.com/ClickHouse/clickhouse-java/blob/c25ffc62c886c0cdc1fcf8d8bf2234690c131dca/clickhouse-data/src/main/java/com/clickhouse/data/ClickHouseCityHash.java

Also cc @zhicwu, it would be good if we can provide some Java version ClickHouse hash function in the official Java client.

Reply:
Yes, I think so. Is it same as in https://github.com/EwainCai/cityHash102-java?

Reply:
It should be the same as https://github.com/EwainCai/cityHash102-java except `ByteBuf` v.s. `byte[]`

Reply:
> We can totally avoid `ByteBuf` in this class, as the `clickhouse-java` does
> 
> https://github.com/ClickHouse/clickhouse-java/blob/c25ffc62c886c0cdc1fcf8d8bf2234690c131dca/clickhouse-data/src/main/java/com/clickhouse/data/ClickHouseCityHash.java
> 
> Also cc @zhicwu, it would be good if we can provide some Java version ClickHouse hash function in the official Java client.

One advantage of using `ByteBuf ` though is if in the future we want to support beyond utf8 string, we can handle type conversion between raw byte arrays and target type by e.g. `ByteBuf.writeDoubleLE`.